### PR TITLE
[RFC]: Functionalize assertions

### PR DESCRIPTION
**Pull request by XuanQi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104203
* #103887
* __->__ #103757
* #103751

The idea here is to create do a graph mutation to:
* Create an initial dependency token at the beginning of the program.
* Replace non-functional version of assertion statements to functional version.
* The functional version of assertion statement will:
  * Accept a dependency token from output of previous functional assertion statement (or the initial dependency token if there isn't any).
  * Generate a dependency token as the output of assertion statement.
  * Augment the output to include the dependency token generated by last assertion statement.

The goal here is to:
* Form an explicit dependency chain and avoid potential reordering during other passes of compiling.
* Make the assertions a part of overall execution graph will affect the final output (or it could potentially be DCEed).

**NOTE:**
* Currently only cover `contrain_range` and WIP to support other assertions. Send out this PR to collect feedback first.
* Here it only focus on implementation itself. Will integrate it with current export in future PR.

### CODE REVIEW

**Comment by avikchaudhuri** [comment]
I have a fundamental design question.

It seems to me that we're using `dep_token`-passing as a means to impose dependency order *only* between stateful ops (e.g., so that they are not re-ordered between themselves).

However, the emphasis is on the word *only*. This feels wrong. It seems to me that the point at which a stateful op is executed is relevant to any other op following it, not just another stateful op.

Consider the following pseudo-code:
```
def f(x):
  assert x > 0
  y = sqrt(x)
  return y
```

Here, it is important that we do not re-order `assert x > 0` and `y = sqrt(x)`.

This can be done in principle by adding a dependency edge from `assert x > 0` to `y = sqrt(x)`. We do not need additional edges from non-stateful ops to non-stateful ops (other than existing dependency edges to encode output/input ordering).

This also hinges on the interpretation that `dep_token` is about preventing re-ordering. On the other hand, it could be purely about avoiding GC (leaving it up to graph transforms to take care and avoid semantically incorrect reordering), in which case I'd argue that we can flatten `dep_token`s: each stateful op returns a token and we return all tokens (or pass them through a special "token reducer" op at the end).

So, in summary:
1. I think `dep_token` doesn't do enough to impose ordering constraints. We can potentially get by with rules such as, push stateful ops as early as possible given their dependencies, and have graph transforms respect that on faith. Or we go further and add dependencies for downstream non-stateful ops on upstream stateful ops as well.
2. We adopt the position that we're just trying to avoid GC and simplify the implementation.

**Comment by avikchaudhuri** [comment]
A good analogy to use when thinking about this is avoiding race conditions in concurrency: here we're accounting for write-write races but not write-read races.

**Comment by XuanQi** [comment]
@avikchaudhuri: so just to verify my understanding to your comment:
> I think dep_token doesn't do enough to impose ordering constraints.

I think what you means here is for this code snippet:
```
assert x > 0
y = sqrt(x)
```
I agree that current dependency strategy is not good enough to prevent compiler re-ordering the above two statements. In this I'm curious what would be a easy way to enforce the ordering here. Would that make sense to instead of returning a dependency token, we could directly return the constrained tensor(s) as well? So the above comes:
```
x , dep_token = functional_assert(x, ...)
y = sqrt(x)
```

>  On the other hand, it could be purely about avoiding GC, ..., I'd argue that we can flatten dep_tokens...

I'm fully get the GC (garbage collection?) part so might need more explanation. Do you actually mean that with this implementation, we're introducing an ordering between assertions which is actually unnecessary? So it would be simply:
```
dep_token_1 = functional_assert(...)
dep_token_2 = functional_assert(...) # No need to accept dep_token_1 as input

return res, reduce_dep_tokens([dep_token_1, dep_token_2, ...])
```

File: torch/_export/functionalize_assertions.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+aten = torch.ops.aten
+
+_NON_FUNCTION_TO_FUNCTIONAL_ASSERTION_FUNCS: Dict[OpOverload, OpOverload] = {
+    aten.sym_constrain_range.default: aten.functional_sym_constrain_range,
+    aten._assert_async.default: aten._functional_assert_async.default,
+    aten._assert_async.msg: aten._functional_assert_async.msg,
+}
+
+
+def functionalize(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:15]: Can we rename it to something like "_functionalize_side_effectful_ops"? There is another public API called `functionalize` in pytorch.

File: torch/_export/functionalize_assertions.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    gm = copy.deepcopy(gm)
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:16]: nit: could you write a simple docstring of what this API roughly does?

File: torch/_export/functionalize_assertions.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import torch
+import copy
+from typing import Dict
+from torch._ops import OpOverload
+
+aten = torch.ops.aten
+
+_NON_FUNCTION_TO_FUNCTIONAL_ASSERTION_FUNCS: Dict[OpOverload, OpOverload] = {
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:8]: nit: _NON_FUNCTIONAL_TO_FUNCTIONAL_SIDE_EFFECTFUL_FUNCS

File: torch/_export/functionalize_assertions.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    gm = copy.deepcopy(gm)
+    graph = gm.graph
+
+    inputs_node = next(n for n in graph.nodes if n.op == "placeholder")
+
+    with graph.inserting_after(inputs_node):
+        dep_token = graph.call_function(torch.empty, args=(0,))
+        dep_token.name = "dep_token"
+
+    none_functional_assertions = [
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:25]: nit: non*

File: torch/_export/functionalize_assertions.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        n
+        for n in graph.nodes
+        if n.target in _NON_FUNCTION_TO_FUNCTIONAL_ASSERTION_FUNCS
+    ]
+    next_dep_token_index = 2
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:30]: Why is it 2?

File: torch/_export/functionalize_assertions.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        with graph.inserting_after(n):
+            kwargs = {**n.kwargs, "dep_token": dep_token}
+            dep_token = graph.call_function(
+                _NON_FUNCTION_TO_FUNCTIONAL_ASSERTION_FUNCS[n.target],
+                args=n.args,
+                kwargs=kwargs,
+            )
+            dep_token.name = f"dep_token_{next_dep_token_index}"
+            next_dep_token_index += 1
+            n.replace_all_uses_with(dep_token)
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:41]: I think you can delete `n` here via graph.erase_node(n)

File: torch/_export/functionalize_assertions.py (hunk 5 above)
Comment (XuanQi) [reply torch/_export/functionalize_assertions.py:30]: So naming of dep tokens are `dep_token`, `dep_token_2` so the next non initial dep token starts at 2. Open to naming suggestion though :p

File: torch/_export/functionalize_assertions.py (hunk 6 above)
Comment (XuanQi) [reply torch/_export/functionalize_assertions.py:41]: Did it at another loop but I guess it's better to just directly erase here.

File: torch/_export/functionalize_assertions.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        dep_token = graph.call_function(aten.make_dep_token)
+        dep_token.name = "dep_token"
+
+    non_functional_assertions = [
+        n
+        for n in graph.nodes
+        if n.target in _NON_FUNCTIONAL_TO_FUNCTIONAL_SIDE_EFFECTFUL_FUNCS
+    ]
+    next_dep_token_index = 2
+    for n in non_functional_assertions:
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:53]: From how it is implemented, I actually think we can reuse the pass infra. (look at AddRuntimeAssertions pass) I initially thought it would be hard to due to creating bunch of new nodes and some carried dependencies, but it doesn't seem too bad. Main benefit is that, once we update the PassInfra with more robust metadata checks, all passes will be updated in one go.

File: torch/_export/functionalize_assertions.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        with graph.inserting_after(n):
+            kwargs = {**n.kwargs, "dep_token": dep_token}
+            dep_token = graph.call_function(
```
Comment (tugsbayasgalan) [inline torch/_export/functionalize_assertions.py:56]: I think we should update/create the node.meta field for this new node. So thats why it is probably if you try to reuse the pass infra.

File: torch/_meta_registrations.py [hunk 9]
```diff
@@ -311,11 +311,34 @@ def assert_async_meta(val, assert_msg):
...
+
+
 @register_meta(aten.sym_constrain_range.default)
 def sym_constrain_range(size, min, max):
     constrain_range(size, min=min, max=max)
 
 
+@register_meta(aten.functional_sym_constrain_range.default)
+def functional_sym_constrain_range(size, min, max, dep_token):
+    constrain_range(size, min=min, max=max)
```
Comment (tugsbayasgalan) [inline torch/_meta_registrations.py:333]: maybe call aten.sym_constrain_range.default instead?

**Comment by avikchaudhuri** [comment]
OK, I've convinced myself that this imperfect dependency tracking is better than the simpler just-avoid-GC-by-returning-all-tokens solution. 

The reasoning is as follows. We want to model effects where 
1. effects depend on values 
2. ordering between effects is important
3. successful computation of values does not depend on effects. 

This is easiest to understand when effects are logging statements instead of assertions: we want to have an ordered log of values but we can take them out and nothing matters.

But it also kinda makes sense for assertions, with the "weak" assumption (3). Previously I was imagining a "stronger" version of (3) where it matters whether we execute `assert x > 0` before `y = sqrt(x)` or after, because we might prefer a failed assertion over a runtime error / computing unexpected values. But on reflection, *successful* computations (where the assertion doesn't fail) are not affected, because the assertion would fail independent of reordering, because it will never be reordered if such reordering means depending on a different `x` (already handled by pure / value dependencies).

File: torch/_export/passes/functionalize_side_effectful_ops_pass.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            return super().call_operator(op, args, kwargs, meta)
+
+        if self._dep_token is None:
+            self._dep_token = super().call_operator(
+                aten._make_dep_token,
+                args=(),
+                kwargs={},
+                meta=self._create_dummy_node_metadata(),
+            )
+            self._dep_token.node.name = "dep_token"
```
Comment (avikchaudhuri) [inline torch/_export/passes/functionalize_side_effectful_ops_pass.py:76]: Pretty sure the convention is to start with `0` or `` here, with the next being `1`. E.g., I've seen both `arg0, arg1, ...` or `add, add1, ...` patterns in graphs.

**Review by avikchaudhuri — APPROVED** [review]
Looks like @tugsbayasgalan 's suggestions were addressed, and my main concern was fundamental. Left a minor nit about numbering scheme of tokens, but code lgtm.
